File: commlibrary/src/main/java/com/anmnight/commlibrary/widget/BitmapShaderRoundImageView.java
package com.anmnight.commlibrary.widget;

import android.content.Context;
import android.graphics.*;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.ImageView;

public class BitmapShaderRoundImageView extends ImageView {

    private Paint mPaint = new Paint();

    public BitmapShaderRoundImageView(Context context) {
        super(context);
    }

    public BitmapShaderRoundImageView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public void onDraw(Canvas canvas) {

        Drawable drawable = getDrawable();

        Bitmap bitmap = scale(((BitmapDrawable) drawable).getBitmap());

        BitmapShader mBitmapShader = new BitmapShader(bitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP);

        mPaint.setShader(mBitmapShader);

        int radius = Math.min(getWidth(), getHeight()) / 2;

        canvas.drawCircle((getWidth() / 2F), (getHeight() / 2F), radius, mPaint);
    }


    private Bitmap scale(Bitmap bitmap) {

        float cWidth = getWidth() - getPaddingLeft() - getPaddingRight();
        float cHeight = getHeight() - getPaddingTop() - getPaddingBottom();
        int oWidth = bitmap.getWidth();
        int oHeight = bitmap.getHeight();

        Matrix matrix = new Matrix();

        float temp = oWidth > oHeight ? cHeight / oHeight : cWidth / oWidth;

        int x = 0;
        int y = 0;
        int len = 0;
        if (oWidth > oHeight) {
            x = (oWidth - oHeight) / 2;
            len = oHeight;
        } else {
            y = (oHeight - oWidth) / 2;
            len = oWidth;
        }

        matrix.postScale(temp, temp);

        return Bitmap.createBitmap(bitmap, x, y, len, len, matrix, false);
    }
}

File: commlibrary/src/main/java/com/anmnight/commlibrary/widget/BitmapShaderRoundImageView.kt
package com.anmnight.commlibrary.widget

import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import androidx.appcompat.widget.AppCompatImageView
import android.util.AttributeSet

class BitmapShaderRoundImageView : AppCompatImageView {

    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    private var mPaint: Paint = Paint()

    override fun onDraw(canvas: Canvas) {

        if (drawable !is BitmapDrawable) {
            throw Exception("drawable isn't BitmapDrawable")
        }

        val bitmap = scale((drawable as BitmapDrawable).bitmap)

        val bitmapShader = BitmapShader(bitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)

        mPaint.shader = bitmapShader

        val radius = Math.min(width,height)/2

        canvas.drawCircle((width / 2).toFloat(), (height / 2).toFloat(), radius.toFloat(), mPaint)
    }


    private fun scale(bitmap: Bitmap): Bitmap {

        val cWidth = width - paddingLeft - paddingRight
        val cHeight = height - paddingTop - paddingBottom

        val oWidth = bitmap.width
        val oHeight = bitmap.height

        val matrix = Matrix()

        val temp = if (oWidth > oHeight) {
            cHeight.toFloat() / oHeight
        } else {
            cWidth.toFloat() / oWidth
        }
        matrix.postScale(temp, temp)

        return Bitmap.createBitmap(bitmap, 0, 0, oWidth, oHeight, matrix, false)
    }
}